### PR TITLE
WIP Update from upstream

### PR DESCRIPTION
**Pull request by Munksgaard**
The goal here is more than to update from upstream, I also want to explore the possibility of using this library as a replacement for the compiletest binary in Rust. It would be nice if there was only one codebase, and it could be used both for testing Rust itself, and for testing external Rust libraries.

Inspired by https://github.com/laumann/compiletest-rs/pull/165#issuecomment-518146972

### CODE REVIEW

File: src/header.rs [hunk 1]
```diff
@@ -212,18 +357,24 @@ pub struct TestProps {
...
-    pub must_compile_successfully: bool,
+    // How far should the test proceed while still passing.
+    pass_mode: Option<PassMode>,
+    // Ignore `--pass` overrides from the command line for this test.
+    ignore_pass: bool,
     // rustdoc will test the output of the `--test` option
     pub check_test_line_numbers_match: bool,
-    // The test must be compiled and run successfully. Only used in UI tests for
-    // now.
-    pub run_pass: bool,
```
Comment (phansch) [inline src/header.rs]: The removal of the two headers has potential to silently change the result of tests. Do you think it makes sense to deprecate them before removing them?

Comment (Munksgaard) [reply src/header.rs]: That's a good point. I wonder why they were removed in the original compiletest.

Comment (phansch) [reply src/header.rs]: IIRC, they were just renamed, not removed. I'm not sure right now which are the correct replacements for each.

Comment (phansch) [reply src/header.rs]: fwiw, I think it would also be fine to add a changelog that explains the changes =)

**Pull request closed without merging**